src/main/java/com/minecolonies/core/datalistener/BaseDataListener.java

```diff
@@ -0,0 +1,221 @@
+package com.minecolonies.core.datalistener;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.*;
+import com.minecolonies.api.util.Log;
+import com.minecolonies.core.util.GsonHelper;
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.JsonOps;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.resources.ResourceManager;
+import net.minecraft.server.packs.resources.SimpleJsonResourceReloadListener;
+import net.minecraft.util.profiling.ProfilerFiller;
+import net.neoforged.neoforge.common.conditions.ICondition;
+import org.apache.commons.lang3.StringUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+
+import static net.neoforged.neoforge.common.conditions.ConditionalOps.DEFAULT_CONDITIONS_KEY;
+
+/**
+ * Base json reload listener class that handles part of the heavy lifting across all reload listeners.
+ *
+ * @param <T> the type of entry.
+ */
+public abstract class BaseDataListener<T> extends SimpleJsonResourceReloadListener
+{
+    /**
+     * JSON keys.
+     */
+    private static final String KEY_REPLACE    = "replace";
+    private static final String KEY_REMOVE     = "remove";
+    private static final String KEY_CONDITIONS = DEFAULT_CONDITIONS_KEY;
+
+    /**
+     * Gson instance
+     */
+    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
+
+    /**
+     * Mapping result class for returning the entry + warnings.
+     *
+     * @param success whether the entry was successfully parsed or not.
+     * @param item    the entry, if successfully parsed.
+     * @param reason  the reason when failed.
+     * @param <T>     the type of entry.
+     */
+    public record MappingResult<T>(
```

Comment:
Don't codecs already have a `DataResult` that does this?

Reply:
Given that Ray wants to stay away as much as possible from Codecs, I opted not to use the full codec way that I initially posted in Discord (it wasn't possible to not use codecs with that way)